parsec/core/local_device.py

```diff
@@ -124,35 +142,40 @@ def slughash(self) -> str:
 
 def list_available_devices(config_dir: Path) -> List[AvailableDevice]:
     try:
-        candidate_pathes = list(get_devices_dir(config_dir).iterdir())
+        key_file_paths = list(get_devices_dir(config_dir).rglob("*.keys"))
     except FileNotFoundError:
         return []
 
     # Sanity checks
     devices = []
-    for device_folder_path in candidate_pathes:
-        slug = device_folder_path.name
-        key_file_path = device_folder_path / f"{slug}.keys"
-
-        try:
-            organization_id, device_id = LocalDevice.load_slug(slug)
-            # Not a valid slug, ignore this folder
-        except ValueError:
-            continue
+    for key_file_path in key_file_paths:
 
         try:
             data = key_file_serializer.loads(key_file_path.read_bytes())
```

Comment:
Wouldn't it be better to have two separate `DeviceFileSchema` classes: `DeviceFileSchema` and `LegacyDeviceFileSchema` (or `DeviceFileSchemaV1` and `DeviceFileSchemaV2`)

this would allow us to have a strong border between old and new formats (which should help to deprecate the old format)
```python
class DeviceFileV2Schema(DeviceFileV1Schema):
    # Store device and organization ID in the device file
    # For legacy versions, this information is available in the file name
    device_id = DeviceIDField(required=True)
    organization_id = OrganizationIDField(required=True)
    root_verify_key_hash = fields.String(required=True)


key_file_v1_serializer = MsgpackSerializer(
    DeviceFileV1Schema, validation_exc=LocalDeviceValidationError, packing_exc=LocalDevicePackingError
)

key_file_v2_serializer = MsgpackSerializer(
    DeviceFileV2Schema, validation_exc=LocalDeviceValidationError, packing_exc=LocalDevicePackingError
)

def load_device_file_v1(key_file_path):
    slug = key_file_path.steam
    try:
        organization_id, device_id = LocalDevice.load_slug(slug)
        # Not a valid slug, ignore this folder
    except ValueError:
        return None

    try:
        data = key_file_v1_serializer.loads(key_file_path.read_bytes())
    except (FileNotFoundError, LocalDeviceError):
        # Not a valid device file, ignore this folder
        return None

    return AvailableDevice(
        key_file_path=key_file_path,
        organization_id=organization_id,
        device_id=device_id,
        human_handle=data["human_handle"],
        device_label=data["device_label"],
    )



def load_device_file_v2():
    try:
        data = key_file_v2_serializer.loads(key_file_path.read_bytes())
    except (FileNotFoundError, LocalDeviceError):
        # Not a valid device file, ignore this folder
        return None

    return AvailableDevice(
        key_file_path=key_file_path,
        organization_id=data["organization_id"],
        device_id=data["device_id"],
        human_handle=data["human_handle"],
        device_label=data["device_label"],
    )

def list_available_devices(config_dir: Path) -> List[AvailableDevice]: 
    [...]
    for key_file_path in key_file_paths:
        for loader in (load_device_file_v1, load_device_file_v2):
            available_device = loader(key_file_path)
            if available_device:
                devices.append(available_device)
```

Reply:
By the way, wouldn't it be better the run a conversion routine before actually retrieving the available devices ?
This would have the benefit of keeping the legacy format handling in a single function (this would solve the issue I was talking about in the `get_key_file` function) and should be really simple to implement given the key file is immutable and all data are available.

On top of that if we don't delete the old file after conversion, we are still compatible with previous version of Parsec  \o/

Reply:
> Wouldn't it be better to have two separate DeviceFileSchema classes: DeviceFileSchema and LegacyDeviceFileSchema (or DeviceFileSchemaV1 and DeviceFileSchemaV2)

I don't really have a strong opinion about it, but there's already a compatibility layer in `DeviceFileSchema`:
```
    # Since human_handle/device_label has been introduced, device_id is
    # redacted (i.e. user_id and device_name are 2 random uuids), hence
    # those fields have been added to the device file so the login page in
    # the GUI can use them to provide useful information.
    human_handle = HumanHandleField(allow_none=True, missing=None)
    device_label = fields.String(allow_none=True, missing=None)
```

Does that mean we'd need `V1`, `V2` and `V3`?  Maybe it's a bit overkill considering the amount of compatibility logic?

Reply:
> By the way, wouldn't it be better the run a conversion routine before actually retrieving the available devices ?
This would have the benefit of keeping the legacy format handling in a single function

Do you mean re-writing the device files from old to new format? Good question :thinking: 

On one hand we would get rid of the old format more easily, but on the other hand writing in the a read function can be a bit dangerous. Plus, we wouldn't be able to fix the `human_handle` and `device_label` fields so we'd need a compatibility layer anyway.

> ... (this would solve the issue I was talking about in the get_key_file function) and should be really simple to implement given the key file is immutable and all data are available.

l used another approach to solve this issue, see 5017400cbeccb8427f9bdab2044dbeabeeade29c

Reply:
Done! 
c809c86